network_simulator/host.py

```diff
@@ -23,4 +32,104 @@ def __init__(self, address=None, flows={}, flow_packets_received={},
         self.flow_packets_received = flow_packets_received
         self.link = link
 
-    # TODO(team): More event subclasses to come?
\ No newline at end of file
+
+class HostReceivedPacketEvent(Event):
+    """
+    Host receives packets from a flow.
+    """
+
+    def __init__(self, host, packet):
+        """
+        Sanity check: make sure destination was this Host and that this isn't
+        a RouterPacket.
+        :ivar host: Host that receives Packet
+        :ivar packet: Packet received
+        :ivar packet_previously_received: True if Packet previously received
+        """
+        Event.__init__(self)
+
+        # Sanity check
+        assert packet.dest_id == host.address and not isinstance(packet,
+                                                                 RouterPacket)
+        self.host = host
+        self.packet = packet
+        self.packet_previously_received = False
+
+    def run(self, main_event_loop, statistics):
+        """
+        Put packet in flow_packets_received and sort.
+        Update statistics.
+        :param main_event_loop: event loop where new Events will be scheduled.
+        :param statistics: statistics to update
+        """
+        # add packet to host.flow_packets_received if its a DataPacket
+        if isinstance(self.packet, DataPacket):
+
+            old_packets_received = self.host.flow_packets_received.get(
+                self.packet.flow_id, default=list())
+
+            self.packet_previously_received = elem_in_list(
+                self.packet.packet_id, old_packets_received)
+
+            if not self.packet_previously_received:
+                bisect.insort(old_packets_received, self.packet.packet_id)
```

Comment:
Add a comment above this line mentioning that this uses O(log(n)) binary search to find the insertion point, and worst-case O(n) insertion.